### ISSUE TITLE
2.1 changes role var precidence

### ISSUE THREAD

**Issue by nirik** [ansible/ansible#16016]
- Bug Report

```
ansible 2.1.0.0
  config file = /home/kevin/.ansible.cfg
  configured module search path = Default w/o overrides
```
##### SUMMARY

In 2.1 ansible uses the role/vars values of variables instead of the passed into the role version. 
##### STEPS TO REPRODUCE

```
Given a playbook: 

- name: ansible 2.1 role variable bug demo
  hosts: thulsadoom.scrye.com
  user: root

  roles:

  - role: testrole
    importantvar: whatipassedtherole

and a role:

% cat roles/testrole/vars/main.yml
importantvar: itstherolevars
% cat roles/testrole/tasks/main.yml

---

# what variable do we have?
- debug: var=importantvar
```

##### EXPECTED RESULTS

With ansible 2.0: 

% ansible-playbook -i inventory -u root playbooks/groups/testing.yml

PLAY [ansible 2.1 role variable bug demo] **************************************

TASK [setup] *******************************************************************
ok: [thulsadoom.scrye.com]

TASK [testrole : debug] ********************************************************
ok: [thulsadoom.scrye.com] => {
    "importantvar": "whatipassedtherole"
}

PLAY RECAP *********************************************************************
thulsadoom.scrye.com       : ok=2    changed=0    unreachable=0    failed=0   
##### ACTUAL RESULTS

```
With 2.1:

PLAY [ansible 2.1 role variable bug demo] **************************************

TASK [setup] *******************************************************************
ok: [thulsadoom.scrye.com]

TASK [testrole : debug] ********************************************************
ok: [thulsadoom.scrye.com] => {
    "importantvar": "itstherolevars"
}

PLAY RECAP *********************************************************************
thulsadoom.scrye.com       : ok=2    changed=0    unreachable=0    failed=0  
```

**Comment by nizig** [comment]
Confirmed. Broke almost every role in our orchestration system.

**Comment by kustodian** [comment]
Even thought this is a bug and it should be fixed, it is not a good practice to override role vars (e.g. `myrole/vars`) outside the role. I would recommend to everyone who has been hit by this to move vars which they need to override from `myrole/vars` to `myrole/defaults`. Role variables should only be used in that role and should never be overridden outside of it.

Also I don't see this being documented in the variable precedence documentation http://docs.ansible.com/ansible/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable

EDIT: It would also be cool, that part of the fix for this a new test is added to automatic tests, so that it can be detected automatically.

**Comment by bcoca** [comment]
just a note, role vars are not ONLY for the role, sometimes you can use a role to provide variable definitions for the rest of the play.

**Comment by phibos** [comment]
Any information if this is a bug or if this is how ansible will handle role vars in upcoming releases?

**Comment by jimi-c** [comment]
Per discussion in our core meeting today (kind of), this appears to be fixed in `stable-2.1` and is included in RC1 for 2.1.1.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
